scipy/optimize/func.py

```diff
@@ -0,0 +1,163 @@
+"""
+Implementations of fitting algorithms for common functions that can be
+computed analytically.
+
+Functions
+---------
+.. autosummary::
+   :toctree: generated/
+
+    exp_fit
+    pow_fit
+"""
+
+from __future__ import division, print_function, absolute_import
+
+__all__ = ['exp_fit', 'pow_fit']
+
+import numpy as np
+from numpy import (
+    argsort, asfarray, cumsum, diff, empty, empty_like, exp, log,
+    square,
+)
+from numpy.linalg import inv
+
+
+def exp_fit(x, y, sorted=True):
+    """
+    Fit an exponential curve to raveled 1D data.
+
+    This algorithm does not require any a-priori knowledge of the data,
+    such as the intercept. The fitting parameters are comptued for:
+
+    .. math::
+
+       y = A + Be^{Cx}
+
+    Parameters
+    ----------
+    x : array-like
+        The x-values of the data points. The fit will be performed on a
+        raveled version of this array.
+    y : array-like
+        The y-values of the data points corresponding to `x`. Must be
+        the same size as `x`. The fit will be performed on a raveled
+        version of this array.
+    sorted : bool
+        Set to True if `x` is already monotonically increasing or
+        decreasing. If False, x will be sorted into increasing order,
+        and y will be sorted along with it.
+
+    Return
+    ------
+    a, b, c : array
+        A 3-element array of optimized fitting parameters. The first
+        element is the additive bias, the second the multiplicative, and
+        the third the exponential.
+
+    Notes
+    -----
+    The fit is computed non-iteratively in a single pass. It can be used
+    to initialize other regression methods based on different
+    optimization criteria. The algorithm and the theory behind it is
+    presented in the paper below.
+
+    References
+    ----------
+    Jacquelin, Jean. “REGRESSIONS Et EQUATIONS INTEGRALES” 14 Jan. 2009, pp. 15–18., https://www.scribd.com/doc/14674814/Regressions-et-equations-integrales
+    """
+    x = asfarray(x).ravel()
+    y = asfarray(y).ravel()
+    if x.size != y.size:
+        raise ValueError('x and y must be the same size')
+    if not sorted:
+        # Is there a better way to do this in scipy?
+        ind = argsort(x)
+        x = x[ind]
+        y = y[ind]
+
+    s = empty_like(y)
+    s[0] = 0
+    s[1:] = cumsum(0.5 * (y[1:] + y[:-1]) * diff(x))
+    # This might be better: Needs a benchmark
```

Comment:
looks like line 81 is faster: https://github.com/jeffyancey/scratchpads/blob/master/exp_fit%20benchmarking.ipynb

```
>>> s[1:] = cumsum(0.5 * (y[1:] + y[:-1]) * diff(x))
The slowest run took 7.19 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 14.3 µs per loop
...
>>> s[1:] = y[:-1]
>>> s[1:] += y[1:]
>>> s[1:] *= np.diff(x)
>>> s = s * .5
>>> s = np.cumsum(s)
he slowest run took 4.89 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 17 µs per loop
```

Reply:
@jeffyancey  And easier to read. Will remove comment. Thanks.